File: c/high-scores/high_scores.c
#include "high_scores.h"

#define TRUE 1
#define FALSE 0

int contains(const size_t *array, size_t array_len, size_t element);
size_t pick_index_of_highest(const int32_t *scores, size_t scores_len,
                             size_t *ignored_indexes,
                             size_t ignored_indexes_len);

int32_t latest(const int32_t *scores, size_t scores_len) {
  return scores[scores_len - 1];
}

int32_t personal_best(const int32_t *scores, size_t scores_len) {
  int32_t best = scores[0];
  for (size_t i = 1; i < scores_len; i++) {
    if (scores[i] > best) {
      best = scores[i];
    }
  }
  return best;
}

size_t personal_top_three(const int32_t *scores, size_t scores_len,
                          int32_t *output) {
  size_t ignored_indexes[2] = {0, 0};

  for (size_t i = 0; i < 3; i++) {
    if (i >= scores_len) {
      return scores_len;
    }
    size_t picked =
        pick_index_of_highest(scores, scores_len, ignored_indexes, i);
    output[i] = scores[picked];
    ignored_indexes[i] = picked;
  }
  return 3;
}

size_t pick_index_of_highest(const int32_t *scores, size_t scores_len,
                             size_t *ignored_indexes,
                             size_t ignored_indexes_len) {
  size_t index_of_highest = scores_len;
  for (size_t i = 0; i < scores_len; i++) {
    if (contains(ignored_indexes, ignored_indexes_len, i)) {
      continue;
    }
    if (index_of_highest == scores_len ||
        scores[i] > scores[index_of_highest]) {
      index_of_highest = i;
    }
  }
  return index_of_highest;
}

int contains(const size_t *array, size_t array_len, size_t element) {
  for (size_t i = 0; i < array_len; i++) {
    if (array[i] == element) {
      return TRUE;
    }
  }
  return FALSE;
}

File: go/high-scores/high_scores.go
// Package highscores provides functionality for managing player scores and retrieving personal bests.
package highscores

// HighScores represents a collection of player scores.
type HighScores struct {
	scores []int
}

// NewHighScores returns a new HighScores instance initialized with the given scores.
func NewHighScores(scores []int) *HighScores {
	return &HighScores{
		scores: scores,
	}
}

// Scores returns all the recorded scores.
func (s *HighScores) Scores() []int {
	return s.scores
}

// Latest returns the latest (last recorded) score.
func (s *HighScores) Latest() int {
	return s.scores[len(s.scores)-1]
}

// PersonalBest returns the best (highest) score.
func (s *HighScores) PersonalBest() int {
	best := s.scores[0]

	for _, v := range s.scores {
		if v > best {
			best = v
		}
	}

	return best
}

// TopThree returns the top three highest scores in descending order.
func (s *HighScores) TopThree() []int {
	topThree := make([]int, 0)

	for _, v := range s.scores {
		topThree = pickBest(v, 3, topThree)
	}

	return topThree
}

// pickBest updates the output slice to contain up to count highest elements, maintaining descending order.
func pickBest(candidate, count int, output []int) []int {
	if len(output) < count {
		output = append(output, candidate)

		return sort(output)
	}

	current := candidate

	for i, v := range output {
		if current > v {
			temp := output[i]
			output[i] = current
			current = temp
		}
	}

	return output
}

// sort sorts a slice of integers in descending order in place.
func sort(vec []int) []int {
	if len(vec) < 2 {
		return vec
	}

	length := len(vec)

	for i := 0; i < length-1; i++ {
		for j := i + 1; j < length; j++ {
			if vec[j] > vec[i] {
				temp := vec[j]
				vec[j] = vec[i]
				vec[i] = temp
			}
		}
	}

	return vec
}
